File: data/visualization/cpp/membership_duration.cpp
#include "matplotlibcpp.h"
#include "csv.h"
#include <vector>
#include <string>
#include <map>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <filesystem>

namespace plt = matplotlibcpp;

std::map<std::string, int> get_data() {
    std::map<std::string, int> membership_counts;
    io::CSVReader<1> in("database/customer.csv");
    in.read_header(io::ignore_extra_column, "member_since");
    std::string member_since;
    while (in.read_row(member_since)) {
        std::tm tm = {};
        std::istringstream ss(member_since);
        ss >> std::get_time(&tm, "%Y-%m-%d");
        std::ostringstream year_month;
        year_month << std::put_time(&tm, "%Y-%m");
        membership_counts[year_month.str()]++;
    }
    return membership_counts;
}

void plot(const std::map<std::string, int>& membership_counts) {
    std::vector<std::string> x;
    std::vector<int> y;
    for (const auto& entry : membership_counts) {
        x.push_back(entry.first);
        y.push_back(entry.second);
    }

    plt::figure_size(1200, 500);
    plt::bar(x, y, 0.5, "skyblue");
    plt::title("Duration of Membership (Monthly)");
    plt::xlabel("Year-Month");
    plt::ylabel("Number of Customers Joined");
    plt::xticks_rotation(45);
    plt::tight_layout();

    std::filesystem::create_directories("images");
    plt::save("images/membership_duration.png");
}

int main() {
    auto membership_counts = get_data();
    plot(membership_counts);

    return 0;
}

File: opti_coffee/visualization/membership_duration.py
import pandas as pd
import matplotlib.pyplot as plt
import os

def get_data():
    # Load the customer data
    df = pd.read_csv('database/customer.csv')

    # Convert 'member_since' column to datetime format
    df['member_since'] = pd.to_datetime(df['member_since'])

    # Extract the year and month from the 'member_since' column
    df['membership_year_month'] = df['member_since'].dt.to_period('M')

    # Count the number of members who joined each month and year
    return df['membership_year_month'].value_counts().sort_index()

def plot(membership_counts):
    plt.figure(figsize=(15,7))
    membership_counts.plot(kind='bar', color='skyblue')
    plt.title('Duration of Membership (Monthly)')
    plt.xlabel('Year-Month')
    plt.ylabel('Number of Customers Joined')
    plt.xticks(rotation=45)
    plt.tight_layout()

    os.makedirs('images', exist_ok=True)
    plt.savefig('images/membership_duration.png')

if __name__ == "__main__":
    membership_counts = get_data()
    plot(membership_counts)